### PR TITLE
SWC-6635: invalidate trashcan cache when entity is deleted

### PR DESCRIPTION
**Pull request by jinjunoh**

### CODE REVIEW

File: src/main/java/org/sagebionetworks/web/client/jsinterop/KeyFactory.java [hunk 1]
```diff
@@ -14,4 +14,6 @@ public KeyFactory(String accessToken) {}
   public native QueryKey getDownloadListBaseQueryKey();
 
   public native QueryKey getEntityQueryKey(String entityId);
+
+  public native QueryKey getTrashCanItemsQueryKey();
```
Comment (jinjunoh) [inline src/main/java/org/sagebionetworks/web/client/jsinterop/KeyFactory.java:18]: Add the getTrashCanItemsQueryKey to KeyFactor to invalidate the cache

File: src/main/java/org/sagebionetworks/web/client/widget/entity/controller/EntityActionControllerImpl.java [hunk 2]
```diff
@@ -2783,6 +2774,14 @@ public void onSuccess(Void result) {
...
             Place gotoPlace = createDeletePlace();
             getGlobalApplicationState().getPlaceChanger().goTo(gotoPlace);
+            KeyFactory keyFactory = keyFactoryProvider.getKeyFactory(
+              authenticationController.getCurrentUserAccessToken()
+            );
+            queryClient.invalidateQueries(
+              InvalidateQueryFilters.create(
+                keyFactory.getTrashCanItemsQueryKey()
+              )
+            );
```
Comment (jinjunoh) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/controller/EntityActionControllerImpl.java:2809]: Use queryClient.invalidateQueries and InvalidateQueryFilters to invalidate the trashcan cache only when deletion is successful

File: src/test/java/org/sagebionetworks/web/unitclient/widget/entity/controller/EntityActionControllerImplTest.java [hunk 3]
```diff
@@ -2600,29 +2854,38 @@ public void testOnDeleteConfirmedPreFlightPassedDeleteFailed() {
...
+    verify(mockView).showErrorMessage(
+      DisplayConstants.ERROR_ENTITY_DELETE_FAILURE + error
+    );
+    QueryKey mockQueryKey = mock(QueryKey.class);
+    when(mockKeyFactory.getTrashCanItemsQueryKey()).thenReturn(mockQueryKey);
+    verify(mockKeyFactoryProvider, never()).getKeyFactory(anyString());
+    verify(mockKeyFactory, never()).getTrashCanItemsQueryKey();
+    verify(mockQueryClient, never()).invalidateQueries(
+      any(InvalidateQueryFilters.class)
+    );
```
Comment (jinjunoh) [inline src/test/java/org/sagebionetworks/web/unitclient/widget/entity/controller/EntityActionControllerImplTest.java]: Test to check that trashcan cache is not invalidated if the deletion is unsuccessful

File: src/test/java/org/sagebionetworks/web/unitclient/widget/entity/controller/EntityActionControllerImplTest.java [hunk 4]
```diff
@@ -2636,19 +2899,30 @@ public void testOnDeleteConfirmedPreFlightPassedDeleteSuccess() {
...
+    verify(mockPlaceChanger).goTo(
+      new Synapse(parentId, null, EntityArea.TABLES, null)
+    );
+    QueryKey mockQueryKey = mock(QueryKey.class);
+    when(mockKeyFactory.getTrashCanItemsQueryKey()).thenReturn(mockQueryKey);
+    verify(mockKeyFactoryProvider).getKeyFactory(anyString());
+    verify(mockKeyFactory).getTrashCanItemsQueryKey();
+    verify(mockQueryClient).invalidateQueries(
+      any(InvalidateQueryFilters.class)
+    );
```
Comment (jinjunoh) [inline src/test/java/org/sagebionetworks/web/unitclient/widget/entity/controller/EntityActionControllerImplTest.java]: Test to check that trashcan cache is invalidated when deletion is successful

**Review by nickgros — APPROVED** [review]
Looks great!

**Pull request merged**